### PR TITLE
Deprecate dynamic mode in OptionParser and promote strict

### LINKED ISSUE #7278: Deprecate dynamic mode in OptionParser and promote strict first

### PR DESCRIPTION
**Pull request by uesteibar**
Fixes #7278

### CODE REVIEW

File: lib/elixir/lib/option_parser.ex [hunk 1]
```diff
@@ -543,6 +545,10 @@ defmodule OptionParser do
           {strict, true}
 
         true ->
+          IO.warn(
+            "Dynamic mode in OptionParser is deprecated, please consider using `strict: [...]`"
```
Comment (josevalim) [inline lib/elixir/lib/option_parser.ex]: I think a better warning would be:

    "Not passing the :switches or :strict option to OptionParser is deprecated".

File: lib/elixir/test/elixir/option_parser_test.exs [hunk 2]
```diff
@@ -400,3 +400,33 @@ defmodule OptionParserTest do
     assert original == OptionParser.to_argv(opts, switches: [counter: :count])
   end
 end
+
+defmodule OptionsParserDeprecationsTest do
```
Comment (josevalim) [inline lib/elixir/test/elixir/option_parser_test.exs]: There are a bunch of warnings now in the OptionParser suite, we need to investigate those too.

File: lib/elixir/lib/option_parser.ex [hunk 3]
```diff
@@ -112,6 +112,8 @@ defmodule OptionParser do
 
   ### Parsing dynamic switches
 
+  WARNING: this option is deprecated.
+
```
Comment (AndrewDryga) [inline lib/elixir/lib/option_parser.ex]: Maybe if the option is deprecated we should delete it from the doc?

File: lib/elixir/test/elixir/option_parser_test.exs (hunk 2 above)
Comment (uesteibar) [reply lib/elixir/test/elixir/option_parser_test.exs]: I'll check them and use either `switch` or `strict` as necessary

Comment (josevalim) [reply lib/elixir/test/elixir/option_parser_test.exs]: One suggestion is to move them all to this module and wrap them with capture_io(:stderr).

**Comment by josevalim** [comment]
Please take a look at the build. It is failing because of unformatted files.

File: lib/elixir/test/elixir/option_parser_test.exs (hunk 2 above)
Comment (josevalim) [reply lib/elixir/test/elixir/option_parser_test.exs]: We already have them in other places in the suite, I think we should rather move them here. :)

**Comment by uesteibar** [comment]
@josevalim addressed your feedback

File: lib/elixir/lib/option_parser.ex (hunk 3 above)
Comment (uesteibar) [reply lib/elixir/lib/option_parser.ex]: I saw it's done like that for `Behaviour`, should we just remove all this @josevalim?

File: lib/elixir/lib/option_parser.ex [hunk 4]
```diff
@@ -543,6 +545,8 @@ defmodule OptionParser do
           {strict, true}
 
         true ->
+          IO.warn("Not passing the :switches or :strict option to OptionParser is deprecated")
+
```
Comment (whatyouhide) [inline lib/elixir/lib/option_parser.ex]: Nitpick: I think we can remove this line.

File: lib/elixir/test/elixir/option_parser_test.exs [hunk 5]
```diff
@@ -400,3 +288,188 @@ defmodule OptionParserTest do
...
+  use ExUnit.Case, async: false
+
+  @warning ~r[Not passing the :switches or :strict option to OptionParser is deprecated]
+
+  defmacro assert_deprecated(clause) do
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/option_parser_test.exs]: I am not convinced we need a macro here. Could we get away with:

```elixir
defp assert_deprecated(fun) do
  assert ExUnit.CaptureIO.capture_io(:stderr, fun) =~ @warning
end
```

and

```elixir
assert_deprecated(fn ->
  assert OptionParser.parse(["--docs"]) == {[docs: true}, [], []}
end)
```

?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**